clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```
@@ -458,10 +462,22 @@ public KafkaProducer(Properties properties, Serializer<K> keySerializer, Seriali
         }
     }
 
-    private static TransactionManager configureTransactionState(ProducerConfig config, LogContext logContext, Logger log) {
+    private static int configureDeliveryTimeout(ProducerConfig config, Logger log) {
+        int deliveryTimeoutMs = config.getInt(ProducerConfig.DELIVERY_TIMEOUT_MS_CONFIG);
+        int lingerMs = config.getInt(ProducerConfig.LINGER_MS_CONFIG);
+        int requestTimeoutMs = config.getInt(ProducerConfig.REQUEST_TIMEOUT_MS_CONFIG);
+        if (deliveryTimeoutMs < Integer.MAX_VALUE && deliveryTimeoutMs < lingerMs + requestTimeoutMs) {
+            // set deliveryTimeoutMs to lingerMs + requestTimeoutMs for backward compatibility
+            deliveryTimeoutMs = lingerMs + requestTimeoutMs;
```

To clarify, I think we should only override the default value. If the user has explicitly provided an inconsistent value, then we should throw an exception. We can check this using `config.originals()`.

---

updated